10_rest/n0npax/pkg/puppy/rest_test.go

```diff
@@ -0,0 +1,226 @@
+package puppy_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	tassert "github.com/stretchr/testify/assert"
+
+	puppy "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy"
+	store "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy/store"
+	"github.com/stretchr/testify/suite"
+)
+
+type storerImpl int
+
+const (
+	syncStorer storerImpl = iota
+	memStorer
+	leveldbStorer
+)
+
+type storerSuite struct {
+	suite.Suite
+	store puppy.Storer
+	impl  storerImpl
+}
+
+func (s *storerSuite) TearDownTest() {
+	if ldbs, ok := s.store.(*store.LevelDBStore); ok {
+		ldbs.Ldb.Close()
+		os.RemoveAll(store.LevelDBPath)
+	}
+}
+
+func (s *storerSuite) SetupTest() {
+	switch s.impl {
+	case syncStorer:
+		s.store = store.NewSyncStore()
+	case memStorer:
+		s.store = store.NewMemStore()
+	case leveldbStorer:
+		store.LevelDBPath = fmt.Sprintf("/tmp/test/%d", time.Now().UnixNano())
+		err := os.Mkdir(store.LevelDBPath, os.ModePerm)
+		if err != nil {
+			s.Fail("LevelDB SetupTest Failed. Cannot create directory")
+		}
+		s.store = store.NewLevelDBStorer()
+	default:
+		panic("Unrecognised storer implementation")
+	}
+
+	for i := 0; i < 5; i++ {
+		p := puppy.Puppy{
+			Breed:  "Type A",
+			Colour: "Grey",
+			Value:  100 * i,
+		}
+		_, err := s.store.CreatePuppy(&p)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func TestStorer(t *testing.T) {
+	suite.Run(t, &storerSuite{impl: syncStorer})
+	suite.Run(t, &storerSuite{impl: memStorer})
+	suite.Run(t, &storerSuite{impl: leveldbStorer})
+}
+
+func performRequest(r http.Handler, method, path string, payload io.Reader) *httptest.ResponseRecorder {
```

Comment:
sendRequest?

Reply:
I usually just split these out into:
```go
 get(h http.Handler, path string)
 post(h http.Handler, path string, payload io.Reader)
 put(h http.Handler, path string, payload io.Reader)
 delete(h http.Handler, path string)
```

Reply:
The other thing I like to do is to use httptest.Server and set it up in setupTest/setup suite.
I use URL as field in suite which I retrieve from server.URL.
By doing so I can use the exact same set of tests and also run them against a compiled standalone server instead of the httptest.Server (possibly even implemented in a different language). All I have to do is pass in the URL of the standalone server as a parameter to the test run. By doing so I effectively get integration tests for free.
I'm not sure if I explained this very clearly - happy to have chat or discuss further here.

Reply:
Point 1 and 2 done.

I believe point 3 is overkill. Instruction doesn't ask us to provide any integration test.
For these labs, I also ignored all optimizations connected with the performance tests, security or metrics.

If this change is required I'll refactor tests.

Reply:
Fair enough - just wanted to mention it because it highlights the advantages of the `httptest.Server` vs `httptest.ResponseRecorder` debate.